### ISSUE TITLE
Improve Trending Tags user experience

### ISSUE THREAD

**Issue by somoso** [mastodon/mastodon#29757]
### Pitch

# Terms

User in this context is admin/moderator/end user who has access to Trending Tags page and permission to change the settings.

# Main

- Add ability for user to search for hashtags, and allow us to filter by review status `approved`, `rejected`, `pending`

Have a search box where the user can search for tags, and maybe also a dropdown dialog that allows filtering by specific review status, or be able to search across all statuses.

- Allow user to manually enter tags for approved/rejected stage

Once a user has selected a review stage, have a text box underneath that allows the user to enter in tags via the text box, and a submit button to allow submitting the manual tag in the list, and assigned to the previously selected review stage automatically.

- Allow user to upload CSV/Text file of approved/rejected tags

Once a user has selected a review stage, have a file upload form underneath that allows the user to upload a file, which said file should have a list of tags. Once submitted, the list of tags will be processed and added automatically to the previously selected review stage automatically.

- Allow user to download CSV/Text file of approved/rejected tags for backup

Once a user has selected a review stage, have a button underneath that allows the user to download the list of tags for the selected review stage. It should download the file in the same format as required by the upload format.

# Extra:

Allow user to enter a hashtag for approved/rejected stage with some sort of pattern matching (could be Regex, could be simpler, could be more complicated)

## Approved scenario:

User has a lot of "X a song" type of hashtags. It would be nice to automatically approve them

## Rejected scenario:

User sees a lot of pornographic hashtags and variants. It would be nice to say `<porn term>*` where `*` is a wildcard item.

### Motivation

As an admin, it is unbelievably frustrating to use this interface to the point where I am tempted to allow all tags to go unmoderated to the detriment of my server and the users on my server - just to save me some time. As an admin, I shouldn't feel like this.

Also, I feel like with the tweaks and suggestions listed, this can also be applied to other moderation areas. Improvements to moderation tools means admins and moderators can do their job efficiently, and provide a safer environment for users of our Mastodon server.

**Comment by ThisIsMissEm** [comment]
> Add ability for user to search for hashtags, and allow us to filter by review status approved, rejected, pending

This has been implemented via the new Hashtags UI in 4.3, see #30880

For the other features here, the question is likely: what happens if the tags haven't been used on your server yet?

> Allow user to download CSV/Text file of approved/rejected tags for backup

For what it's worth, the total number of hashtags on a server can be extremely high, so I'm not sure this would necessarily make sense. e.g., on Hachyderm.io, we currently have 3.6 million known hashtags; and there are 43.5k hashtag follows, of which 16k are for unique hashtags. We've also 3.6 million trendable tags, but only 681 tags that are manually approved for trending.

**Comment by Justinzobel** [comment]
I'd love to be able to sort the tags by how many people it's been used by over the last week. It would make it easier to see which are used and likely need review.

**Comment by ThisIsMissEm** [comment]
> I'd love to be able to sort the tags by how many people it's been used by over the last week. It would make it easier to see which are used and likely need review.

I think that exists only for trending hashtags since that value iirc has to be calculated & cached

**Comment by Justinzobel** [comment]
Yes, that's what I want, only in the trending hash tags.

**Comment by ThisIsMissEm** [comment]
@Justinzobel see Trending -> Hashtags, I think that already exists

**Comment by Justinzobel** [comment]
Ah I'm looking at `https://server/admin/trends/tags?status=pending_review`

**Comment by ThisIsMissEm** [comment]
Yeah, if memory serves that's sorted by recency. I'd have to look into whether a different mode is possible. Can you ping on this issue in November? I'm currently on sick leave until then

**Comment by ThisIsMissEm** [comment]
Could you also retest on 4.3, since we added a stable sort order to trending hashtags pending review in 4.3

**Comment by Justinzobel** [comment]
@RemindMe in 30 days to revisit this page!
